### DISCUSSION TITLE
Erratic Flight using LPS

### DISCUSSION THREAD

**Discussion by magicbycalvin** [bitcraze/discussions#51] [Q&A] [answered]
Hello,

I have been having issues with erratic flight using the Loco Positioning System (LPS). I followed all the steps mentioned on the Bitcraze website for the setup of the Crazyflie 2.1 and the LPS. The Crazyflie's firmware was updated to the most recent release as of today and the firmware for the LPS nodes was updated to the most recent release as of yesterday. I have also looked through previous discussions with similar issues but haven't found an acceptable solution.

When attempting to fly using the take off command within the `cfclient`, the Crazyflie will either simply sit on the ground and spin some of its motors a little bit, but not get off the ground, or it will fly incredibly erratically (see the attached video) 

https://user-images.githubusercontent.com/17996040/170614744-68f7967a-87f7-47af-961f-2e7b6d7bb5a5.mp4

I also encounter similar problems when attempting to run some of the basic Python examples. However, I can fly the Crazyflie without an issue using a gamepad. I should note that I have successfully gotten the LPS system to work at home a year or two ago, but now it will not work in my lab. The thing that confuses me the most is that the setup I had at my home was far from ideal, while the setup at my lab very closely resembles the suggested setup on the Bitcraze website. I do not have any reason to believe that there are significant sources of noise at my lab that would interfere with the UWB beacons.

Other potentially relevant information, in no particular order:

- Operating system: Ubuntu 20.04
- Number of LPS beacons: 8
- LPS mode: in the video I am using TWR but I have had the same experience using TDoA2 and TDoA3
- I have had the same flight experience using at least three different Crazyflies and I have access to a total of 20 I can test
- Batteries have all been v

erified to be full before flight
- When looking at the LPS tab, the position estimate bounces around within the expected +/- 10 cm tolerance but it will sometimes drift much further than it should, oftentimes resulting in the state estimation error being printed to the console

I am attaching the logs from my last flight (shown in the video) and for convenience have also created some plots below. Any help will be much appreciated.

Thank you very much for your time!

![LoPoSys](https://user-images.githubusercontent.com/17996040/170612137-83885909-271c-48dd-8085-cc28b133ef10.png)
![LoPoTab0](https://user-images.githubusercontent.com/17996040/170612140-6945b714-03a9-4a92-a47f-bcf1c0a17a10.png)
![LoPoTab1](https://user-images.githubusercontent.com/17996040/170612143-bec560dd-4178-423b-bb2f-1071ce6e684f.png)
![Motors](https://user-images.githubusercontent.com/17996040/170612146-0a291071-da78-410f-b33e-7d10c0ff8a5b.png)
![Pose](https://user-images.githubusercontent.com/17996040/170612147-f5e07069-fe9b-4cb7-b7b7-1a5e8e227777.png)
![stabilizer](https://user-images.githubusercontent.com/17996040/170612148-2c029a6a-4a2d-4da8-a8ff-06feb4baad65.png)
![Battery](https://user-images.githubusercontent.com/17996040/170612150-7599d9b7-6c4e-4cb0-bf52-f2e0a2566bda.png)

[Pose-20220526T16-45-43.csv](https://github.com/bitcraze/discussions/files/8783327/Pose-20220526T16-45-43.csv)
[stabilizer-20220526T16-45-44.csv](https://github.com/bitcraze/discussions/files/8783328/stabilizer-20220526T16-45-44.csv)
[Battery-20220526T16-45-40.csv](https://github.com/bitcraze/discussions/files/8783332/Battery-20220526T16-45-40.csv)
[LoPoSys-20220526T16-45-41.csv](https://github.com/bitcraze/discussions/files/8783333/LoPoSys-20220526T16-45-41.csv)
[LoPoTab0-20220526T16-45-41.csv](https://github.com/bitcraze/discussions/files/8783334/LoPoTab0-20220526T16-45-41.csv)
[LoPoTab1-20220526T16-45-42.csv](https://github.com/bitcraze/discussions/files/8783335/LoPoTab1-20220526T16-45-42.csv)
[Motors-20220526T16-45-43.csv](https://github.com/bitcraze/discussions/files/8783336/Motors-20220526T16-45-43.csv)

**Comment by magicbycalvin** [comment, marked as answer]
Well, it looks like I missed something. I'll keep this post up in hopes of helping others in the future!

It turns out I missed the fact that when using the LPS, the Crazyflie must be facing the positive X direction. Instead, I made the mistake of starting the Crazyflie in the positive Y direction which is why the system would go unstable.

**Reply by knmcguire** [reply]
Ah yes a common mistake... it's in the tutorial but we have to make it super big and bold, and even then the most experiences fliers make the mistake after a while, myself included 😁.

But we need to have a localization conversion check during flight...

**Reply by magicbycalvin** [reply]
I am also curious whether it would be possible to self localize the UWB beacons. I wouldn't be opposed to looking into the math side of things if you folks haven't already!

**Reply by knmcguire** [reply]
Ah yeah its an interesting optimization problem but not implemented yet. Could you start a new thread for this, in 'ideas'? Then I can share some publications next week that might be interesting for this

**Reply by fida-121** [reply]
Thanks man I was getting this error for three days and did not find any solution. It was my first time working with LPS. It was great help.

**Comment by MINJEEgi** [comment]
Hello, I'm not sure if it's possible to receive a response to an old post, but the post mentioned that the crazyflie needs to face the x direction. How do you determine the absolute x direction inside a room? Is it determined by the placement of anchors?

**Reply by magicbycalvin** [reply]
That is correct. The placement of the anchors is what decides where your X, Y, and Z axes are. When you set up the anchors, you need to manually specify the XYZ position of each one. In doing so, you effectively choose the location and orientation of the origin. I found it helpful to use masking tape on the ground to specify which directions are X and Y to help avoid confusion when setting up.

**Reply by eunseo01** [reply]
how we decide xyz position in manually? the placement of the anchor automatically decide xyz?

**Comment by eunseo01** [comment]
how we decide xyz position in manually? the placement of the anchor automatically decide  xyz?

**Reply by krichardsson** [reply]
yes
